src/Compilers/Shared/BuildServerConnection.cs

```diff
@@ -595,4 +621,205 @@ public static string GetTempPath(string workingDir)
+        void Unlock ();
+        bool IsDisposed { get; }
+    }
+
+    internal sealed class FileMutex : IDisposable
+    {
+        public readonly bool OwnsMutex;
+        public readonly FileStream Stream;
+        public readonly string FilePath;
+
+        public bool IsLocked { get; private set; }
+
+        internal static string GetMutexDirectory ()
+        {
+            var homeDirectory = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
+            if (String.IsNullOrWhiteSpace(homeDirectory))
+                homeDirectory = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
+            if (String.IsNullOrWhiteSpace(homeDirectory))
+                homeDirectory = Path.GetTempPath();
+            var result = Path.Combine(homeDirectory, ".roslyn");
+            if (!Directory.Exists(result))
+                Directory.CreateDirectory(result);
+            return result;
+        }
+
+        public FileMutex (string name)
+        {
+            FilePath = Path.Combine(GetMutexDirectory(), name);
+            try
+            {
+                Stream = new FileStream(FilePath, FileMode.Open, FileAccess.ReadWrite, FileShare.None);
+                OwnsMutex = false;
+            }
+            catch (FileNotFoundException)
+            {
+                Stream = new FileStream(FilePath, FileMode.CreateNew, FileAccess.ReadWrite, FileShare.None);
+                OwnsMutex = true;
+            }
+        }
+
+        public bool TryLock (int timeoutMs)
+        {
+            if (IsLocked)
+                throw new InvalidOperationException("Lock already held");
+
+            var sw = Stopwatch.StartNew();
+            do
+            {
+                try
+                {
+                    Stream.Lock(0, 0);
+                    IsLocked = true;
+                    return true;
+                }
+                catch (IOException)
+                {
+                    // Lock currently held by someone else
+                    Thread.Sleep(0);
+                }
+                catch (Exception)
+                {
+                    // Something else went wrong.
+                    return false;
+                }
+            } while (sw.ElapsedMilliseconds < timeoutMs);
+
+            return false;
+        }
+
+        public void Unlock ()
+        {
+            if (!IsLocked)
+                return;
+            Stream.Unlock(0, 0);
+            IsLocked = false;
+        }
+
+        public void Dispose ()
+        {
+            var wasLocked = IsLocked;
+            if (wasLocked)
+                Unlock();
+            Stream.Dispose();
+            // We do not delete the lock file here because there is no reliable way to perform a
+            //  'delete if no one has the file open' operation atomically on *nix. This is a leak.
+        }
+    }
+
+    internal sealed class ServerNamedMutex : IServerMutex
+    {
+        private readonly Mutex Mutex;
```

Comment:
The naming convention for `private` fields is prefix with _ and camel cased. For this though I think you could also just make it a `public` property as well. It's reasonabel to expose this.